### PR TITLE
Set cpu limits if there is a resource quota with limits

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
Currently, there is a mechanism by whereby if there is a Resource Quota with memory
limits and the `AutoResourceLimits` FG is enabled, Kubevirt will automatically
set the memory limits in the virt-launcher pod.

With CPU limits, another behavior has been implemented:
define a namespace label selector(autoCPULimitNamespaceLabelSelector) that instructs in which namespace the automatic cpu
limits will be set.

The same functionality of the memory limits can be applied to the cpu limits too, so
that if the AutoResourceLimits` FG is enabled, Kubevirt will automatically
set both memory and cpu limits in the virt-launcher pod if a RQ with limits is associated
to the creation namespace.

`autoCPULimitNamespaceLabelSelector` struct will extend the functionality if the admin decides
to enable this on namespaces where RQ is not present.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (https://github.com/kubevirt/user-guide/pull/722).
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
Automatically set cpu limits when a resource quota with cpu limits is associated to the creation namespace and the `AutoResourceLimits` FeatureGate is enabled
```

/cc @vladikr @jean-edouard @xpivarc @acardace

### CODE REVIEW

**Comment by Barakmor1** [comment]
Hey @fossedihelm ,
What if the resourceQuota was created after the vmi was scheduled?
Wouldn't future migration fail because of this?
Don't you think the user should configure if limits should be added instead of detecting resourcesQuotas in the namespace?

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -1355,7 +1354,12 @@ func checkForKeepLauncherAfterFailure(vmi *v1.VirtualMachineInstance) bool {
 	return keepLauncherAfterFailure
 }
 
-func requiresCPULimits(virtClient kubecli.KubevirtClient, labelSelector *metav1.LabelSelector, vmi *v1.VirtualMachineInstance) bool {
+func (t *templateService) doesVMIRequireAutoCPULimits(vmi *v1.VirtualMachineInstance) bool {
+	if namespaceWithRQ := t.doesVMIRequireAutoResourceLimits(vmi, k8sv1.ResourceCPU); namespaceWithRQ {
```
Comment (jean-edouard) [inline pkg/virt-controller/services/template.go:1358]:
```suggestion
	if t.doesVMIRequireAutoResourceLimits(vmi, k8sv1.ResourceCPU) {
```
Unless I'm missing something?

File: pkg/virt-controller/services/template.go [hunk 2]
```diff
@@ -1365,13 +1369,27 @@ func requiresCPULimits(virtClient kubecli.KubevirtClient, labelSelector *metav1.
 		log.DefaultLogger().Reason(err).Warning("invalid CPULimitNamespaceLabelSelector set, assuming none")
 		return false
 	}
-	namespace, err := virtClient.CoreV1().Namespaces().Get(context.Background(), vmi.Namespace, metav1.GetOptions{})
+
+	if t.namespaceStore == nil {
+		return false
```
Comment (jean-edouard) [inline pkg/virt-controller/services/template.go:1374]: Should we log an error?

**Review by jean-edouard — COMMENTED** [review]
/approve
Great improvement!
Couple minor comments in-line.
Please make sure to update the user guide once this PR merges:
https://kubevirt.io/user-guide/virtual_machines/resources_requests_and_limits/

File: pkg/virt-controller/services/template.go (hunk 1 above)
Comment (fossedihelm) [reply pkg/virt-controller/services/template.go]: Thank you!!! You are definitely right

**Comment by fossedihelm** [comment]
> /approve Great improvement! Couple minor comments in-line. Please make sure to update the user guide once this PR merges: https://kubevirt.io/user-guide/virtual_machines/resources_requests_and_limits/

Thanks @jean-edouard for reminding me to update the user guide. I will work on it right away.

**Pull request merged**